### PR TITLE
Handles/improved overlapping visualization + mouse movement + compile under 2019.2

### PR DESCRIPTION
**Pull request by LogicalError**
When moving the mouse cursor over the horizon, it would switch to movement over a camera aligned plane to avoid making the movement going into infinity.
Unfortunately when calculating the intersection point on this alternative plane, it would rotate it incorrectly, losing one axis of movement.
I fixed this by ensuring that the shared axis between the two planes is always the same when rotating the point from one plane to another

### CODE REVIEW

File: Packages/com.chisel.components/Chisel/Components/API.private/ChiselGeneratedComponentManager.cs [hunk 1]
```diff
@@ -65,8 +65,13 @@ public static bool NeedUVGeneration(ChiselModel model)
                 return false;
 
             var staticFlags = UnityEditor.GameObjectUtility.GetStaticEditorFlags(model.gameObject);
+#if UNITY_2019_2_OR_NEWER
+            if ((staticFlags & UnityEditor.StaticEditorFlags.ContributeGI) != UnityEditor.StaticEditorFlags.ContributeGI)
+                return false;
+#else
```
Comment (LogicalError) [inline Packages/com.chisel.components/Chisel/Components/API.private/ChiselGeneratedComponentManager.cs]: flag has changed between 2019.1 and 2019.2

File: Packages/com.chisel.editor/Chisel/Editor/Editor/ChiselUnityEventsManager.cs [hunk 2]
```diff
@@ -155,6 +155,11 @@ private static void GridOnSceneGUI(SceneView sceneView)
                     grid = UnitySceneExtensions.Grid.ActiveGrid;
                 grid.Render(sceneView);
             }
+
+            if (UnitySceneExtensions.Grid.debugGrid != null)
+            {
+                UnitySceneExtensions.Grid.debugGrid.Render(sceneView);
+            }
```
Comment (LogicalError) [inline Packages/com.chisel.editor/Chisel/Editor/Editor/ChiselUnityEventsManager.cs]: Added ability to render a debug grid, useful when debugging moving over a plane

File: Packages/com.chisel.editor/Chisel/Editor/Editor/Scene/Handles/PointDrawing.cs [hunk 3]
```diff
@@ -98,6 +104,8 @@ public static void Reset()
 
             if (s_StartIntersection != null)
             {
+                if (!dragArea.Contains(mousePosition))
+                    return null;
```
Comment (LogicalError) [inline Packages/com.chisel.editor/Chisel/Editor/Editor/Scene/Handles/PointDrawing.cs]: Tried to make drawing a shape that is larger than your sceneview work, but something weird is happening on the side of unity. Need to investigate this before I can turn it back on. For now just ignore any mouse movement outside the sceneview

File: Packages/com.chisel.editor/Chisel/Editor/Editor/Scene/Handles/PointDrawing.cs [hunk 4]
```diff
@@ -111,11 +119,14 @@ static void UpdatePoints(List<Vector3> points, Vector3? point)
         {
             if (point.HasValue)
             {
+                var localPosition = s_InvTransform.MultiplyPoint(point.Value);
+                if (s_CurrentPointIndex > 0 &&
+                    (localPosition - points[s_CurrentPointIndex - 1]).sqrMagnitude < kDistanceEpsilon)
+                    return;
```
Comment (LogicalError) [inline Packages/com.chisel.editor/Chisel/Editor/Editor/Scene/Handles/PointDrawing.cs]: Ignore any point that's the same as the last point when doing free-draw/rectangle drawing

File: Packages/com.chisel.editor/Chisel/Editor/Editor/Scene/Handles/PointDrawing.cs [hunk 5]
```diff
@@ -127,27 +138,48 @@ static void UpdatePoints(List<Vector3> points, Vector3? point)
...
+            GUIUtility.hotControl = 0;
+            GUIUtility.keyboardControl = 0;
+            EditorGUIUtility.SetWantsMouseJumping(0);
+            s_MousePosition = Event.current.mousePosition;
+            s_MouseJumping = false;
+        }
+
         static void Commit(Event evt, Rect dragArea, ref List<Vector3> points)
         {
+            var newPoint = GetPointAtPosition(s_MousePosition, dragArea);
```
Comment (LogicalError) [inline Packages/com.chisel.editor/Chisel/Editor/Editor/Scene/Handles/PointDrawing.cs]: If we fail to add the last point, for whatever reason, we cancel

File: Packages/com.chisel.editor/Chisel/Editor/Editor/Scene/Handles/PointDrawing.cs [hunk 6]
```diff
@@ -203,6 +235,20 @@ public static void Do(int id, Rect dragArea, ref List<Vector3> points, out Matri
...
+                            using (new UnityEditor.Handles.DrawingScope(s_Transform))
+                            {
+                                var count = points.Count - 1;
+                                for (int i = 0; i < count - 1; i++)
+                                {
+                                    if ((points[count] - points[i]).sqrMagnitude < kDistanceEpsilon)
+                                    {
+                                        if (i > 0)
+                                            UnityEditor.Handles.color = Color.red;
+                                        capFunction(-1, points[count], orientation, UnityEditor.HandleUtility.GetHandleSize(points[count]) * (kPointScale * 2.0f), type);
```
Comment (LogicalError) [inline Packages/com.chisel.editor/Chisel/Editor/Editor/Scene/Handles/PointDrawing.cs]: Render a 'highlight' of sorts when the current point is hovering another point, make it red (invalid) when it's not the first point (which will close the loop)

File: Packages/com.chisel.editor/Chisel/Editor/Editor/Scene/Handles/PointDrawing.cs [hunk 7]
```diff
@@ -214,7 +260,15 @@ public static void Do(int id, Rect dragArea, ref List<Vector3> points, out Matri
                         GUIUtility.hotControl != id)
                         break;
 
-                    UpdatePoints(points, GetPointAtPosition(evt.mousePosition, dragArea));
+
+                    if (s_MouseJumping)
+                        s_MousePosition += evt.delta;
+                    else
+                        s_MousePosition = Event.current.mousePosition;
```
Comment (LogicalError) [inline Packages/com.chisel.editor/Chisel/Editor/Editor/Scene/Handles/PointDrawing.cs]: When SetWantsMouseJumping is set to 1 (true) then the mouse can loop around the screen. But this only works if we keep track of the relative mouse position versus the current absolute mouse position

File: Packages/com.chisel.editor/Chisel/Editor/Editor/Scene/Handles/PointDrawing.cs [hunk 8]
```diff
@@ -233,7 +295,9 @@ public static void Do(int id, Rect dragArea, ref List<Vector3> points, out Matri
                     if (SceneHandles.InCameraOrbitMode)
                         break;
 
-                    if (GUIUtility.hotControl != 0)
+                    var hotControl = GUIUtility.hotControl;
+                    if (hotControl != 0 &&
+                        hotControl != id)
```
Comment (LogicalError) [inline Packages/com.chisel.editor/Chisel/Editor/Editor/Scene/Handles/PointDrawing.cs]: When releaseOnMouseUp is false, then the current hotControl might still point to our own id.
After we clicked the first time, it won't be set to 0. But we still want to be able to click again

File: Packages/com.chisel.editor/Chisel/Editor/Editor/Scene/Handles/ShapeExtrusionHandle.cs [hunk 9]
```diff
@@ -54,7 +54,7 @@ static Curve2D GetShape()
 
             var pointCount = s_ExtrusionMode ? s_Points.Count - 1 : s_Points.Count;
 
-            if ((s_Points[pointCount - 1] - s_Points[0]).sqrMagnitude < kDistanceEpsilon)
+            if ((s_Points[pointCount - 1] - s_Points[0]).sqrMagnitude < PointDrawing.kDistanceEpsilon)
```
Comment (LogicalError) [inline Packages/com.chisel.editor/Chisel/Editor/Editor/Scene/Handles/ShapeExtrusionHandle.cs]: To ensure we have the same epsilon value in ShapeExtrusionHandle and PointDrawing (which shows the highlight) I've moved this to PointDrawing.

File: Packages/com.scene.handle.extensions/Handle Extensions/Editor/Snapping/SceneHandles.Snapping.cs [hunk 10]
```diff
@@ -1,3 +1,4 @@
+//#define ENABLE_DEBUG_GRID
```
Comment (LogicalError) [inline Packages/com.scene.handle.extensions/Handle Extensions/Editor/Snapping/SceneHandles.Snapping.cs]: When uncommented, a debug grid is shown when moving something over a plane

File: Packages/com.scene.handle.extensions/Handle Extensions/Editor/Snapping/SceneHandles.Snapping.cs [hunk 11]
```diff
@@ -533,18 +537,41 @@ public void CalculateExtents(Vector3[] localPoints)
...
+            var tangent         = Vector3.Cross(worldSlidePlane.normal, normal);
+            if (tangent.sqrMagnitude < 0.0001f)
+            {
+                worldPlanePosition = worldSlideOrigin;
+                return false;
+            }
+            tangent.Normalize();
+            var fromQuaternion  = Quaternion.LookRotation(tangent, normal);
+            var toQuaternion    = Quaternion.LookRotation(tangent, worldSlidePlane.normal);
+            var rotation        = fromQuaternion * Quaternion.Inverse(toQuaternion);
```
Comment (LogicalError) [inline Packages/com.scene.handle.extensions/Handle Extensions/Editor/Snapping/SceneHandles.Snapping.cs]: To ensure we only rotate along one axis (the axis where the two planes intersect) we take the cross product between the two planes and make sure that when we rotate, this axis stays the same

File: Packages/com.scene.handle.extensions/Handle Extensions/Editor/Snapping/SceneHandles.Snapping.cs [hunk 12]
```diff
@@ -562,17 +589,28 @@ private bool GetPlaneIntersection(Vector2 guiPosition, out Vector3 worldPlanePos
                 var origin = worldSlidePlane.ClosestPointOnPlane(worldSlideOrigin);
                 return GetIntersectionOnAlternativePlane(worldRay, normal, origin, out worldPlanePosition);
             }
-
+            
             if (!worldSlidePlane.Raycast(worldRay, out dist)) { dist = float.PositiveInfinity; }
-            if (dist > camera.farClipPlane)
+
+            float farClipPlaneDistance = camera.farClipPlane * 0.5f;
```
Comment (LogicalError) [inline Packages/com.scene.handle.extensions/Handle Extensions/Editor/Snapping/SceneHandles.Snapping.cs]: Not sure if 0.5f is a good number, but farClipPlane is definitely too far away to place the alternative plane on.
Might want to revisit this solution later on

**Comment by otdavies** [comment]
I appreciate your commenting on your solutions and thinking here. @LogicalError 

Reading through everything, it seems solid.

**Pull request merged**